### PR TITLE
Implementation of Olaf

### PR DESCRIPTION
**Pull request by Fiono11**
This PR builds upon the previous PR #102 and implements the [Olaf](https://eprint.iacr.org/2023/899.pdf) protocol. It is part of deliverable 1 of Web3 Foundation's grant https://github.com/w3f/Grants-Program/pull/2250.

### CODE REVIEW

File: .rustfmt.toml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+trailing_comma = "Vertical"
+# trailing_semicolon = false
+# use_field_init_shorthand = true
+
+# where_single_line = true  # does not work on fn
+# brace_style = "AlwaysNextLine"  # does not work on method fn
+
+# Format comments
+comment_width = 100
+wrap_comments = true
```
Comment (burdges) [inline .rustfmt.toml:30]: Your git GUI has seemingly pulled recent changes from master into its own commit.  I'd expect git handles this fine, likely disapears once rebased onto master, but it's odd behavior that indicates issues with the git GUI.

Did you make any other changes to this PR since 25 April?

File: Cargo.toml [hunk 2]
```diff
@@ -22,17 +22,25 @@ curve25519-dalek = { version = "4.1.0", default-features = false, features = [
     "zeroize",
     "precomputed-tables",
     "legacy_compatibility",
+    "rand_core",
+    "serde",
```
Comment (burdges) [inline Cargo.toml:26]: How is rand_core being added as a feature?  rand_core is not an ptional dependency

File: .rustfmt.toml (hunk 1 above)
Comment (Fiono11) [reply .rustfmt.toml:30]: Yeah, I messed up...Pushed some changes directly to master instead of this branch and then pulled them.

File: Cargo.toml (hunk 2 above)
Comment (Fiono11) [reply Cargo.toml:26]: I am not sure, does it have to be?

Comment (burdges) [reply Cargo.toml:26]: rand_core is mandatory.  why is this feature being added?

Comment (Fiono11) [reply Cargo.toml:26]: Because without it it does not compile, I don't know why.

**Comment by Fiono11** [comment]
Replaced by #107.
